programs/validator-history/src/state.rs

```diff
@@ -364,6 +409,65 @@ impl ValidatorHistory {
         Ok(())
     }
 
+    /// Given epoch credits from the vote account, determines which entries do not exist in the history and inserts them.
+    /// Shifts all existing entries that come later in the history and evicts the oldest entries if the buffer is full.
+    /// Skips entries which are not already in the (min_epoch, max_epoch) range of the buffer.
+    pub fn insert_missing_entries(
+        &mut self,
+        epoch_credits: &[(
+            u64, /* epoch */
+            u64, /* epoch cumulative votes */
+            u64, /* prev epoch cumulative votes */
+        )],
+    ) -> Result<()> {
+        // For each epoch in the list, insert a new entry if it doesn't exist
+        let start_epoch = cast_epoch(
+            epoch_credits
+                .iter()
+                .min_by_key(|(epoch, _, _)| *epoch)
+                .unwrap()
+                .0,
+        )?;
+
+        let end_epoch = cast_epoch(
+            epoch_credits
+                .iter()
+                .max_by_key(|(epoch, _, _)| *epoch)
+                .unwrap()
+                .0,
+        )?;
+
+        let entries = self
```

Comment:
idk if this collect is needed, can you iter/enumerate in the for loop below?

Reply:
forgot - it's actually necessary cause this is an immutable borrow of self.history and then we mutably borrow when inserting the missing entries, so need to make a copy. will be fine cause it's a vec of bools which gets allocated on heap